### PR TITLE
ODE Solver for transient and nonlinear processes – organ transplantation

### PR DESCRIPTION
**Pull request by chleh**
TODO
* [x] cleanup
* [x] docs
* [x] adapt input files

Future work:
* save some matrices (i.e. save space)
* adaptive time stepping
* different error norms
* numerical Jacobians
* process coupling
* ...

### CODE REVIEW

File: Applications/ApplicationsLib/ProjectData.cpp [hunk 1]
```diff
@@ -40,8 +45,54 @@ void readGeometry(std::string const& fname, GeoLib::GEOObjects & geo_objects)
 	gml_reader.readFile(fname);
 }
 
+// TODO copied from BoostXmlGmlInterface.cpp
+//! Method for handling conversion to string uniformly across all types and std::string; see std::string overload below.
+template<typename T> std::string tostring(T const& value)
```
Comment (chleh) [inline Applications/ApplicationsLib/ProjectData.cpp]: TODO put to baselib

File: Applications/ApplicationsLib/ProjectData.cpp [hunk 2]
```diff
@@ -57,9 +108,11 @@ ProjectData::ProjectData(BaseLib::ConfigTree const& project_config,
 		);
 
 	MeshLib::Mesh* const mesh = FileIO::readMeshFromFile(mesh_file);
-	if (!mesh)
+	if (!mesh) {
 		ERR("Could not read mesh from \'%s\' file. No mesh added.",
 			mesh_file.c_str());
+		std::abort();
```
Comment (chleh) [inline Applications/ApplicationsLib/ProjectData.cpp]: TODO problem for GUI?

File: Applications/ApplicationsLib/ProjectData.cpp [hunk 3]
```diff
@@ -134,6 +191,44 @@ bool ProjectData::removeMesh(const std::string &name)
...
 }
 
+void ProjectData::buildProcesses()
+{
+	for (auto const& pc : _process_configs)
+	{
+		auto const type = pc.peekConfParam<std::string>("type");
+
+		auto const nl_slv_name = pc.getConfParam<std::string>("nonlinear_solver");
+		auto& nl_slv_pair = detail::get_or_error(_nonlinear_solvers, nl_slv_name,
```
Comment (chleh) [inline Applications/ApplicationsLib/ProjectData.cpp]: TODO make not pair!

File: Applications/ApplicationsLib/UncoupledProcessesTimeLoop.h [hunk 4]
```diff
@@ -0,0 +1,329 @@
...
+        PerProcessData(NumLib::NonlinearSolver<Matrix, Vector, NLTag>& nl_slv,
+                       TimeDisc& time_disc,
+                       NumLib::ODESystem<Matrix, Vector, ODETag, NLTag>& ode_sys
+                       )
+            : nonlinear_solver_tag(NLTag)
+            , nonlinear_solver(nl_slv)
+            , tdisc_ode_sys(
+                  new NumLib::TimeDiscretizedODESystem<Matrix, Vector, ODETag, NLTag>(
+                                ode_sys, time_disc))
+            , mat_strg(dynamic_cast<NumLib::InternalMatrixStorage&>(*tdisc_ode_sys))
```
Comment (chleh) [inline Applications/ApplicationsLib/UncoupledProcessesTimeLoop.h]: TODO discuss in detail

File: Applications/ApplicationsLib/UncoupledProcessesTimeLoop.h [hunk 5]
```diff
@@ -0,0 +1,329 @@
...
+            std::abort();
+    }
+}
+
+
+template<typename Matrix, typename Vector>
+void
+UncoupledProcessesTimeLoop<Matrix, Vector>::
+initInternalData(ProjectData& project,
+                 std::vector<PerProcessData>& time_disc_ode_syss,
```
Comment (chleh) [inline Applications/ApplicationsLib/UncoupledProcessesTimeLoop.h]: TODO return it, not pass it.

File: Applications/ApplicationsLib/UncoupledProcessesTimeLoop.h [hunk 6]
```diff
@@ -0,0 +1,329 @@
...
+        using NLSlvPicard = NumLib::NonlinearSolver<Matrix, Vector, Tag::Picard>;
+        using NLSlvNewton = NumLib::NonlinearSolver<Matrix, Vector, Tag::Newton>;
+
+        if (auto* nlslv_picard = dynamic_cast<NLSlvPicard*>(&nl_solver))
+        {
+            // The Picard solver can also work with a Newton-ready ODE,
+            // because the Newton ODESystem derives from the Picard ODESystem.
+            // So no further checks are needed here.
+
+            return { *nlslv_picard, time_disc, ode_sys };
```
Comment (chleh) [inline Applications/ApplicationsLib/UncoupledProcessesTimeLoop.h]: add `PerProcessData`

File: Applications/ApplicationsLib/UncoupledProcessesTimeLoop.h [hunk 7]
```diff
@@ -0,0 +1,329 @@
...
+        auto& pcs = **p;
+        auto& time_disc = pcs.getTimeDiscretization();
+        auto& ode_sys   = *time_disc_ode_syss[pcs_idx].tdisc_ode_sys;
+        auto const nl_tag = time_disc_ode_syss[pcs_idx].nonlinear_solver_tag;
+
+        auto const num_eqs = ode_sys.getNumEquations();
+
+        // TODO maybe more is required for PETSc
+        // append a solution vector of size num_eqs
+        process_solutions.emplace_back(num_eqs);
```
Comment (chleh) [inline Applications/ApplicationsLib/UncoupledProcessesTimeLoop.h]: TODO return it, not pass it

File: Applications/ApplicationsLib/UncoupledProcessesTimeLoop.h [hunk 8]
```diff
@@ -0,0 +1,329 @@
...
+setInitialConditions(ProjectData& project,
+                     double const t0,
+                     std::vector<PerProcessData>& time_disc_ode_syss,
+                     std::vector<Vector>& process_solutions)
+{
+    unsigned pcs_idx = 0;
+    for (auto p = project.processesBegin(); p != project.processesEnd();
+         ++p, ++pcs_idx)
+    {
+        auto& pcs = **p;
```
Comment (chleh) [inline Applications/ApplicationsLib/UncoupledProcessesTimeLoop.h]: TODO unify

File: AssemblerLib/VectorMatrixAssembler.h [hunk 9]
```diff
@@ -12,39 +12,35 @@
...
+ *  \c GlobalMatrix'es and \c GlobalVector's.
+ *
+ * It optionally gets the local d.o.f. from a GlobalVector using
+ * a LocalToGlobalIndexMap and passes them on to the local assembler.
+ *
+ * Each type of equation as flagged by the \c NLTag will have a different
+ * VectorMatrixAssembler type.
+ */
+template<typename GlobalMatrix, typename GlobalVector,
+         NumLib::ODESystemTag NLTag>
```
Comment (chleh) [inline AssemblerLib/VectorMatrixAssembler.h]: TODO ODETag

File: ProcessLib/Process.h [hunk 10]
```diff
@@ -87,61 +106,144 @@ class Process
...
+		return *_time_discretization;
 	}
 
 protected:
-	virtual void post(){};
-
-	/// Set linear solver options; called by the derived process which is
-	/// parsing the configuration.
-	void setLinearSolverOptions(BaseLib::ConfigTree&& config)
+	virtual void post(GlobalVector const& x)
```
Comment (chleh) [inline ProcessLib/Process.h]: remove (void)

File: ProcessLib/GroundwaterFlowFEM.h [hunk 11]
```diff
@@ -91,8 +101,7 @@ class LocalAssemblerData : public LocalAssemblerDataInterface<GlobalMatrix, Glob
         _localRhs.reset(new NodalVectorType(local_matrix_size));
     }
 
-    void assemble(std::vector<double> const& /*local_x*/,
-                  std::vector<double> const& /*local_x_prev_ts*/) override
+    void assemble(double const, std::vector<double> const&) override
```
Comment (chleh) [inline ProcessLib/GroundwaterFlowFEM.h]: add commented names

File: ProcessLib/GroundwaterFlowFEM.h [hunk 12]
```diff
@@ -109,13 +118,12 @@ class LocalAssemblerData : public LocalAssemblerDataInterface<GlobalMatrix, Glob
         }
     }
 
-    void addToGlobal(
-        GlobalMatrix& A, GlobalVector& rhs,
-        AssemblerLib::LocalToGlobalIndexMap::RowColumnIndices const& indices)
+    void addToGlobal(AssemblerLib::LocalToGlobalIndexMap::RowColumnIndices const& indices,
+        GlobalMatrix& /*M*/, GlobalMatrix& K, GlobalVector& b)
```
Comment (chleh) [inline ProcessLib/GroundwaterFlowFEM.h]: todo override

File: NumLib/ODESolver/NonlinearSolver.h [hunk 13]
```diff
@@ -50,33 +93,23 @@ class NonlinearSolver<Matrix, Vector, NonlinearSolverTag::Newton> final
...
-     *
-     * \note This method is needed to preload CrankNicolson time discretization scheme.
-     *       It is not used for the general solver steps; in those only the solve() method
-     *       is needed.
-     *
-     * \param sys the equation system to be assembled.
-     * \param x   the state at which the equation system will be assembled.
-     */
-    void assemble(System& sys, Vector const& x) const;
+    void setEquationSystem(System& eq) { _equation_system = &eq; }
```
Comment (chleh) [inline NumLib/ODESolver/NonlinearSolver.h]: TODO maybe 1:1 correspondence process -- nonlinear solver

**Comment by chleh** [comment]
As far as I can see, the PETSc benchmarks fail (but maybe you also want to check). They will be fixed in a subsequent PR.

**Comment by chleh** [comment]
All points from monday's discussion have been addressed.

**Comment by wenqing** [comment]
@chleh  Great! I am going to review it in the next week.

File: Applications/ApplicationsLib/UncoupledProcessesTimeLoop.h [hunk 14]
```diff
@@ -0,0 +1,365 @@
...
+     *
+     * \note Currently the \c ODETag is automatically inferred from the given
+     *       \c ody_sys. This works as long as \c Process derives from
+     *       \c ODESystem<Matrix, Vector, ODETag>, i.e. as long we only deal with
+     *       one type of ODE. When we introduce more types, this method will have
+     *       to be extended slightly.
+     */
+    template<NumLib::ODESystemTag ODETag>
+    static
+    PerProcessData
```
Comment (wenqing) [inline Applications/ApplicationsLib/UncoupledProcessesTimeLoop.h]: A suggestion: PerProcessData -->SingleProcessData

File: Applications/ApplicationsLib/UncoupledProcessesTimeLoop.h [hunk 15]
```diff
@@ -0,0 +1,365 @@
...
+            setEquationSystem<Tag::Picard>(nl_solver, eq_sys);
+            break;
+        case Tag::Newton:
+            setEquationSystem<Tag::Newton>(nl_solver, eq_sys);
+            break;
+        }
+    }
+};
+
+//! Build an UncoupledProcessesTimeLoop from the given configuration.
```
Comment (wenqing) [inline Applications/ApplicationsLib/UncoupledProcessesTimeLoop.h]: In consistent with the previous writing style, Build --> Builds.

File: MathLib/LinAlg/LinearSolver.cpp [hunk 16]
```diff
@@ -0,0 +1,105 @@
...
+{
+    using Solver = typename DefaultLinearSolver<Matrix, Vector>::Solver;
+    return createLinearSolver<Matrix, Vector, Solver>(config);
+}
+
+} // MathLib
+
+
+// specializes the DefaultLinearSolver<> class
+#define DfltLinSlv(MAT, VEC, SLV) \
```
Comment (wenqing) [inline MathLib/LinAlg/LinearSolver.cpp]: Suggestion:  *Slv ---> *Solver

**Comment by wenqing** [comment]
@chleh :+1:

File: MathLib/LinAlg/LinearSolver.cpp (hunk 16 above)
Comment (endJunction) [inline MathLib/LinAlg/LinearSolver.cpp]: I'm also suggesting to write out all of the "short cuts". The time to recognize DfltLinSlv is (for me) definitely longer than for DefaultLinearSolver.

Macros should be upper case.

File: Applications/ApplicationsLib/UncoupledProcessesTimeLoop.h (hunk 7 above)
Comment (endJunction) [inline Applications/ApplicationsLib/UncoupledProcessesTimeLoop.h]: Could you push the newest version, please?

File: Applications/ApplicationsLib/ProjectData.cpp [hunk 17]
```diff
@@ -134,6 +147,44 @@ bool ProjectData::removeMesh(const std::string &name)
...
+		else
+		{
+			ERR("Unknown process type: %s", type.c_str());
+			std::abort();
+		}
+	}
+
+	// process configs are not needed anymore, so clear the storage
+	// in order to trigger config tree checks
+	_process_configs.clear();
```
Comment (endJunction) [inline Applications/ApplicationsLib/ProjectData.cpp]: Better to pass the configs as a unique_ptr to this function...

File: Applications/ApplicationsLib/UncoupledProcessesTimeLoop.h [hunk 18]
```diff
@@ -0,0 +1,365 @@
...
+
+    struct PerProcessData
+    {
+        template<NumLib::ODESystemTag ODETag, NumLib::NonlinearSolverTag NLTag>
+        PerProcessData(NumLib::NonlinearSolver<Matrix, Vector, NLTag>& nl_slv,
+                       TimeDisc& time_disc,
+                       NumLib::ODESystem<Matrix, Vector, ODETag, NLTag>& ode_sys
+                       )
+            : nonlinear_solver_tag(NLTag)
+            , nonlinear_solver(nl_slv)
```
Comment (endJunction) [inline Applications/ApplicationsLib/UncoupledProcessesTimeLoop.h]: I also do like the style where the alignment happens on the colon/comma but until now the standard is to use alignment on the word begin. At the very least it should be consistent in a single file.

File: Applications/ApplicationsLib/UncoupledProcessesTimeLoop.h [hunk 19]
```diff
@@ -0,0 +1,365 @@
...
+    using EquationSystem = NumLib::EquationSystem;
+    //! An abstract process
+    using Process        = ProcessLib::Process<GlobalSetupType>;
+    //! An abstract time discretization
+    using TimeDisc       = NumLib::TimeDiscretization<Vector>;
+
+    struct PerProcessData
+    {
+        template<NumLib::ODESystemTag ODETag, NumLib::NonlinearSolverTag NLTag>
+        PerProcessData(NumLib::NonlinearSolver<Matrix, Vector, NLTag>& nl_slv,
```
Comment (endJunction) [inline Applications/ApplicationsLib/UncoupledProcessesTimeLoop.h]: There are three abbreviations of the nonlinear solver objects: nl_slv, nl_solver, nlslv_... .
Please choose one. Be consistent. I prefer nonlinear_solver. (autocomplete?)

File: Applications/ApplicationsLib/UncoupledProcessesTimeLoop.h [hunk 20]
```diff
@@ -0,0 +1,365 @@
...
+
+    //! Constructs and returns a \c vector of PerProcessData from the given
+    //! ProjectData.
+    std::vector<PerProcessData> initInternalData(ProjectData& project);
+
+    //! Sets and returns initial conditions for the given ProjectData and
+    //! PerProcessData.
+    std::vector<Vector> setInitialConditions(
+            ProjectData& project, double const t0,
+            std::vector<PerProcessData>& per_process_datas);
```
Comment (endJunction) [inline Applications/ApplicationsLib/UncoupledProcessesTimeLoop.h]: C'mon! What is datas? vector of childs?
per_process_data is something which is holding data per process already.
Rename PerProcessData -> ProcessData, keep per_process_data.

File: AssemblerLib/VectorMatrixAssembler.h [hunk 21]
```diff
@@ -68,30 +66,70 @@ class VectorMatrixAssembler
...
+            indices.insert(indices.end(), idcs.begin(), idcs.end());
         }
 
         LocalToGlobalIndexMap::RowColumnIndices const r_c_indices(
                     indices, indices);
 
-        local_assembler->assemble(localX, localX_pts);
-        local_assembler->addToGlobal(_A, _rhs, r_c_indices);
+        local_assembler->assemble(t);
+        local_assembler->addToGlobal(r_c_indices, b);
```
Comment (endJunction) [inline AssemblerLib/VectorMatrixAssembler.h]: Understanding question: For the first order implicit quasi linear system a system Mdx/dt+Kx+b=0 is assembled. What equation is assembled here for the "neumann bc" specialization?
Maybe this will help to choose a better name instead of NeumannBC...

File: BaseLib/uniqueInsert.h [hunk 22]
```diff
@@ -17,6 +17,8 @@
```
Comment (endJunction) [inline BaseLib/uniqueInsert.h]: FYI: this file and the string tools could be own PRs.

File: BaseLib/uniqueInsert.h [hunk 23]
```diff
@@ -26,6 +28,56 @@ void uniquePushBack(Container& container, typename Container::value_type const&
...
+{
+	auto it = map.find(key);
+	if (it == map.end()) {
+		ERR("%s Key `%s' does not exist.", error_message.c_str(), tostring(key).c_str());
+		std::abort();
+	}
+
+	return it->second;
+}
+//! \overload
```
Comment (endJunction) [inline BaseLib/uniqueInsert.h]: overload for const maps.

File: MathLib/LinAlg/LinearSolver.h [hunk 24]
```diff
@@ -0,0 +1,59 @@
+#pragma once
```
Comment (endJunction) [inline MathLib/LinAlg/LinearSolver.h]: header missing. ifdef MATHLIB_LINALG_LINEARSOLVER_H_ ...

File: MathLib/LinAlg/LinearSolver.cpp [hunk 25]
```diff
@@ -0,0 +1,105 @@
+#include "LinearSolver.h"
```
Comment (endJunction) [inline MathLib/LinAlg/LinearSolver.cpp]: header missing

File: MathLib/LinAlg/PETSc/PETScLinearSolver.h [hunk 26]
```diff
@@ -34,23 +35,24 @@ namespace MathLib
 /*!
      A class of linear solver based on PETSc rountines.
 
+     TODO does that still hold?
```
Comment (endJunction) [inline MathLib/LinAlg/PETSc/PETScLinearSolver.h]: Addendum: All unrecognized (by ogs) options are passed to the linear solver. The linear solver options in the project file do have precedence because they are parsed later than the command line options.

File: NumLib/ODESolver/TimeLoopSingleODE.h [hunk 27]
```diff
@@ -29,17 +29,22 @@ template<typename Matrix, typename Vector, NonlinearSolverTag NLTag>
...
      * \param nonlinear_solver The solver to be used to resolve nonlinearities.
      */
     explicit
-    TimeLoopSingleODE(TDiscODESys& ode_sys, NLSolver& nonlinear_solver)
+    TimeLoopSingleODE(TDiscODESys& ode_sys,
+                      std::unique_ptr<LinearSolver>&& linear_solver,
+                      std::unique_ptr<NLSolver>&& nonlinear_solver)
         : _ode_sys(ode_sys)
-        , _nonlinear_solver(nonlinear_solver)
+        , _linear_solver   (std::move(linear_solver))
```
Comment (endJunction) [inline NumLib/ODESolver/TimeLoopSingleODE.h]: alignment on word boundaries...

File: ProcessLib/GroundwaterFlowFEM.h [hunk 28]
```diff
@@ -34,11 +40,15 @@ class LocalAssemblerDataInterface
...
-                          std::vector<double> const& local_x_prev_ts) = 0;
+    virtual void assemble(double const t, std::vector<double> const& local_x) = 0;
 
-    virtual void addToGlobal(GlobalMatrix& A, GlobalVector& rhs,
-            AssemblerLib::LocalToGlobalIndexMap::RowColumnIndices const&) const = 0;
+    // TODO: Don't we waste quite some time by having this as a virtual method?
+    //       If we put a callback arg to the assemble() call, this funcitonality can
+    //       be provided by a non-virtual method.
+    //       Otherwise we could also pass local matrices in with the assemble() call.
+    //       Thereby we would not have to expose global matrices/vectors.
```
Comment (endJunction) [inline ProcessLib/GroundwaterFlowFEM.h]: Suggest to move such discussions into issues, because here in the code nobody will see it.

File: ProcessLib/GroundwaterFlowProcess.h [hunk 29]
```diff
@@ -82,10 +101,13 @@ class GroundwaterFlowProcess : public Process<GlobalSetup>
                 this->_integration_order);
     }
 
+    // TODO remove, but put "gw_" somewhere
```
Comment (endJunction) [inline ProcessLib/GroundwaterFlowProcess.h]: Is this fixed in the followup PR with PETSc fixes?

File: ProcessLib/GroundwaterFlowProcess.h [hunk 30]
```diff
@@ -124,6 +143,20 @@ class GroundwaterFlowProcess : public Process<GlobalSetup>
     using LocalAssembler = GroundwaterFlow::LocalAssemblerDataInterface<
         typename GlobalSetup::MatrixType, typename GlobalSetup::VectorType>;
 
+
+    void assembleConcreteProcess(const double t, GlobalVector const& x,
+                                 GlobalMatrix& M, GlobalMatrix& K, GlobalVector& b) override
+    {
+        // TODO It looks like, with little work this entire method can be moved to the Process class.
```
Comment (endJunction) [inline ProcessLib/GroundwaterFlowProcess.h]: Note taken.

**Comment by endJunction** [comment]
Please rebase and squash. If possible extract the small pieces of unrelated fixes.

File: Applications/ApplicationsLib/ProjectData.cpp (hunk 17 above)
Comment (chleh) [inline Applications/ApplicationsLib/ProjectData.cpp]: This line has not been introduced in this PR, rather it has been moved here from the header. Will fix it in a subsequent PR.

File: Applications/ApplicationsLib/UncoupledProcessesTimeLoop.h (hunk 18 above)
Comment (chleh) [inline Applications/ApplicationsLib/UncoupledProcessesTimeLoop.h]: I always align initializer lists like this and method argument lists at word boundaries. So it's consistent. ^^
I don't want to change my style here.

File: Applications/ApplicationsLib/UncoupledProcessesTimeLoop.h (hunk 19 above)
Comment (chleh) [inline Applications/ApplicationsLib/UncoupledProcessesTimeLoop.h]: I apologize. That has been knitted with a hot needle. ^^

File: Applications/ApplicationsLib/UncoupledProcessesTimeLoop.h (hunk 20 above)
Comment (chleh) [inline Applications/ApplicationsLib/UncoupledProcessesTimeLoop.h]: Misuse of language, but "s" is indicating "List of things".

File: AssemblerLib/VectorMatrixAssembler.h (hunk 21 above)
Comment (chleh) [inline AssemblerLib/VectorMatrixAssembler.h]: Let's discuss vis-a-vis.

File: BaseLib/uniqueInsert.h (hunk 23 above)
Comment (chleh) [inline BaseLib/uniqueInsert.h]: Suggest to keep it, s.t. the method is readily usable also for the const case.

File: MathLib/LinAlg/LinearSolver.cpp (hunk 25 above)
Comment (chleh) [inline MathLib/LinAlg/LinearSolver.cpp]: How could that happen...

File: NumLib/ODESolver/TimeLoopSingleODE.h (hunk 27 above)
Comment (chleh) [inline NumLib/ODESolver/TimeLoopSingleODE.h]: I oppose :no_entry:

File: ProcessLib/GroundwaterFlowProcess.h (hunk 29 above)
Comment (chleh) [inline ProcessLib/GroundwaterFlowProcess.h]: Not yet, but I guess, there will be a discussion about the PETSc PR.

**Comment by chleh** [comment]
Jenkins, test this please.

**Comment by chleh** [comment]
Only PETSc tests are still failing. They will be fixed in #1060.
